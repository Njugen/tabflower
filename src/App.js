import React, { Fragment, Component } from 'react';


/* Import Modals */

import CalendarDateSettingsModal from "./components/modals/calendarDateSettingsModal";
import ConfirmationModal from "./components/modals/confirmationModal";

/* Import Routes */
import RouteList from './components/routes/routeList';

/* Import Sidebars */
import MainSidebar from './components/sidebars/mainSidebar/mainSidebar';

/* Import Utilities */
import FullWidthLoadbar from './components/utils/fullWidthLoadbar';

/* Import Footer */
import ViewFooter from './components/views/components/viewFooter';



/* Import CSS generated by create-react-app */
import './styles/react-generated/App.css';

/* Import Tabeon app specific CSS */
import './styles/tabeon/style.css';
import { ReactDOM } from 'react-dom';



class App extends Component{
  state = {
    currentView: {},
    routes: [],
    modal: {},
    mainSidebar: {},
    refreshFactor: 0
  };

 
  updateState = (newProps, showLoadbar, callback) => {
    /* 
      Used to update the App state itself. This function works the same way as this.setState(), 
      but also increases the refreshFactor (the FullWidthLoadbar uses this to determine whether to launch the loadbar
        or not)
    */
    let newState = {
      ...newProps
    }

    if(showLoadbar && showLoadbar === true){
      let { refreshFactor } = this.state;
      refreshFactor++;

      newState.refreshFactor = refreshFactor;
    }

    this.setState(newState, callback);
  }
  
  handleNavigation = (viewProps) => {

    this.updateState(
      {
        currentView: viewProps
      },
      true
    )
  }

  handleMainSidebarClick = (sidebarProps) => {
    this.updateState(
      {
        mainSidebar: sidebarProps,
      
      },
      false
    )
  }

  modalHandler = (data) => {
    console.log("BATMAN", data);
    if(data.clear && data.clear === true){
      this.clearModal();
    } else {
      this.launchModal(data);
    }
  }

  launchModal = (data) => {
    const modal = {
      launched: true,
      ...data
    }

    this.setState({
      modal
    });
  }

  clearModal = () => {
    const modal = {};

    setTimeout(() => {
      this.setState({ modal }); 
    }, 500)
    
  }
  
  handleRouteListReady = (data) => {
    const routes = data;

    this.setState({ routes });
  }
  

  render = () => {
    const { launched: modalLaunched, id: modalId } = this.state.modal;
    
    return (
      <Fragment>
       
        <div className="container-fluid">
          {(modalLaunched && modalId === "confirm-action") && <ConfirmationModal data={this.state.modal} onSave={() => ""} onDismiss={() => this.clearModal()}></ConfirmationModal>}
          {(modalLaunched && modalId === "date-settings") && <CalendarDateSettingsModal data={this.state.modal} onSave={() => ""} onDismiss={() => this.clearModal()}></CalendarDateSettingsModal>}
          <div className="row">
            <MainSidebar routes={this.state.routes} onMainSidebarClick={(data) => this.handleMainSidebarClick(data)} />
            <div className="col-10 py-2" id="tabeon-view-container">
                <RouteList onRaisedRoutesInfo={(data) => this.handleRouteListReady(data)} onRaiseToModal={(data) => this.modalHandler(data)} onNavigation={(data) => this.handleNavigation(data)} />
                <ViewFooter />
            </div>
          </div>
          
        </div>
        <FullWidthLoadbar refreshFactor={this.state.refreshFactor} />
      </Fragment>
    );
  }

}

export default App;
