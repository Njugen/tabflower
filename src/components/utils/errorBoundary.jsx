import React, {Component} from "react";
import ErrorOverlay from '../modals/errorOverlay';
import { PropTypes } from 'prop-types';

/*
  Catch errors in child component's lifecycle hooks, during rendering and in 
  component constructors. For each error caught, show a fallback UI.

  Read more about error boundaries in React:
  https://reactjs.org/docs/error-boundaries.html
*/

class ErrorBoundary extends Component {
    state = { errors: [] };

     /* Update state so the next render will show the fallback UI. */
    static getDerivedStateFromError(error) {

      let currentErrors = [];
      const newError = error;
      currentErrors.push(newError);

      return { errors: currentErrors };
    }
    
    /*
      Catch an error and add it as an object in the this.state.error array for later use.
    */
    componentDidCatch(error, errorInfo) {
        let currentErrors = this.state.errors;
        const newError = Object.assign(error, errorInfo);
        currentErrors.push(newError);
        
        this.setState({
            errors: currentErrors
        }, () => {
            
        })
    }
  
    render() {
      if (this.state.errors.length > 0) {
        /*
          If errors are stored in the boundary state, render the ErrorOverlay UI component and inform the user about the errors
          using the information stored in this.state.errors
        */
        return <ErrorOverlay data={this.state.errors} onSave={() => ""} onDismiss={() => window.location.reload()}></ErrorOverlay>
      }
  
      return this.props.children || null; 
    }
  }

  ErrorBoundary.propTypes = {
    children: PropTypes.any.isRequired
  }

  export default ErrorBoundary;