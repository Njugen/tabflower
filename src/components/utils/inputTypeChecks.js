const isEmptyString = (input) => {
    if(typeof input === "string"){
        if(input.length === 0 || input.replace(/ /g, "").length === 0){
            return true;
        } else {
            false
        }
    } else {
        return false;
    }
}