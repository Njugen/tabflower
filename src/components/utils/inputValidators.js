/*
    isObject()

    Check if the input is an object, 
    input = {}

    Params:
    - input (can be whatever)

    Returns:
    - true, if the input is an object
    - false, if the input is not an object
*/

const isObject = (input) => {
    if(typeof input === "object"){
        return true;
    } else {
        return false;
    }
} 

/*
    isArray

    Check if the input is an array.

    Params:
    - input (can be whatever)

    Returns:
    - true, if input is an array
    - false, if input is not an array or if input is not defined
*/

const isArray = (input) => {
    if(typeof input !== "undefined"){
        return Array.isArray(input);
    } else {
        return false;
    }
}

/*
    isEmptyString()

    Check if a value is an empty string. A value is an empty
    string if all of the fllowing conditions are met:
        - The input is a string
        - The input's length is 0 OR
        - The input only consists of empty space

    Params:
    - input (can be whatever)

    Returns: 
        - true, if the input is an empty string
        - false, if the input is not an empty string
*/

const isEmptyString = (input) => {
    if(typeof input === "string"){
        if(input.length === 0 || input.replace(/ /g, "").length === 0){
            return true;
        } else {
            return false
        }
    } else {
        return false;
    }
}

/*
    isString()

    Check if the input value is a string, which consists of more than 0 characters (not empty).
    
    Params:
    - input (can be whatever)
    
    Returns:
    - true, if input is string and longer than 0 characters
    - false, if the input is not a string OR if the input is a string consisting of 0 characters
*/

const isString = (input) => {
    if(typeof input === "string"){
        if(input.length > 0){
            return true;
        } else {
            return false
        }
    } else {
        return false;
    }
}

/*
    isNumber()

    Verify that the input is actually a number

    Parameters:
    - input (Can be whatever)

    Returns:
    - true, if the input is a number (any number)
    - false, if the input is not a number
*/
const isNumber = (input) => {
    if(typeof input === "number"){
        return true;
    } else {
        return false
    }
}

/*
    isPositiveNumber()

    Check whether the input is a positive number or not

    Params:
    - input (can be anything)

    Returns:
    - true, if the input is a number AND the input is larger than 0
    - false, if the input is a number but not larger than 0
    - false, if the input is not a number
*/
const isPositiveNumber = (input) => {
    if(typeof input === "number"){
        if(input > 0){
            return true;
        } else {
            return false;
        }
    } else {
        return false;
    }
}

/*
    isNegativeNumber()

    Check if the input is a negative number

    Params:
    - input (can be any value)

    Returns:
    - true, if the input is a number and less than 0
    - false, if the input is a number, but larger than 0
    - false, if the input is not a number
*/
const isNegativeNumber = (input) => {
    if(typeof input === "number"){
        if(input < 0){
            return true;
        } else {
            return false;
        }
    } else {
        return false;
    }
}

