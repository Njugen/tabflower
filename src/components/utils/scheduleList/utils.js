const getWeekdaysArray = () => {
  let weekdays = [];

  weekdays[0] = "sunday";
  weekdays[1] = "monday";
  weekdays[2] = "tuesday";
  weekdays[3] = "wednesday";
  weekdays[4] = "thursday";
  weekdays[5] = "friday";
  weekdays[6] = "saturday";

  return weekdays;
};

export const getTodayObj = () => {
  const date = new Date();
  const weekdays = getWeekdaysArray();

  const day = date.getDay();

  return {
    dayName: weekdays[day],
    order: day,
  };
};

export const getUpcomingDayObj = (dayName) => {
  const weekdays = getWeekdaysArray();

  return {
    dayName,
    order: weekdays.indexOf(dayName),
  };
};

export const determineDayDifference = (today, futureDay) => {
  if (today > 6) return false;

  let n = 0;

  if (today >= futureDay) {
    n += 7;
  } else {
    n = 0;
  }

  const diff = today !== futureDay ? (futureDay + n - today) % 7 : 7;

  return diff;
};

export const determineFutureDate = (numOfDaysFromNow) => {
  const currentDateAsMS = Date.now();
  const numOfDaysAsMS = 60 * 60 * 24 * 1000 * numOfDaysFromNow;

  const futureDate = new Date(currentDateAsMS + numOfDaysAsMS);

  const startOfFutureDate = new Date(
    futureDate.getFullYear(),
    futureDate.getMonth(),
    futureDate.getDate(),
    0,
    0,
    0
  );

  return {
    date: startOfFutureDate.getDate(),
    month: startOfFutureDate.getMonth(),
    year: startOfFutureDate.getFullYear(),
    startOfDateTimeInMS: startOfFutureDate.getTime(),
  };
};

export const convertTimeToMS = (hour, minute) => {
  const hourParamInMS = hour * 60 * 60 * 1000;
  const minuteParamInMS = minute * 60 * 1000;

  return hourParamInMS + minuteParamInMS;
};
