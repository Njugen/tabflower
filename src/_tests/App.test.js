import React from 'react';
import { shallow } from 'enzyme';
import App from './../App';
import * as ExceptionsHandler from './../components/utils/exceptionsAndHandler';
import * as validator from './../components/utils/inputValidators';


const predefinedComponent = () => {
    const component = shallow(<App />);
    return component;
}

let testComponent;
let componentInstance;

describe("Test <App /> component behaviour at mount", () => {
    
    beforeEach(() => {
        testComponent = predefinedComponent();
    });

    test("The state is correctly preset according to template with no modifications, when no functions change the state at mount", () => {
        const componentInstance = testComponent.instance();
        const templateState = {
            currentView: {},
            routes: [],
            modal: {},
            errors: [],
            MainNavBar: {},
            refreshFactor: 0
        };
        
        expect(componentInstance.state).toStrictEqual(templateState);
    });
});

describe("Test functions in <App />", () => {
    let ExceptionsHandler = jest.requireActual("./../components/utils/exceptionsAndHandler");

    // Copy our error messages from exceptionsAndHandler.js. These messages are to be mocked.
    const actualErrorReturns = {
        "app-101": ExceptionsHandler.ValidatorError("app-101"),
        "app-102": ExceptionsHandler.ValidatorError("app-102"),
        "app-103": ExceptionsHandler.ValidatorError("app-103"),
        "app-104": ExceptionsHandler.ValidatorError("app-104"),
        "app-105": ExceptionsHandler.ValidatorError("app-105"),
        "app-106": ExceptionsHandler.ValidatorError("app-106")
    };

    // The error messages we expect to catch and be used as parameter in componentInstance.launchErrorOverlay, if an error has occured
    const expectedErrorReturns = {
        "app-101": {
            name: "ValidatorError",
            message: "The \"callback\" parameter in the App component's updateState() function needs to be a function, or undefined.",
            code: "app-101"
        },
        "app-102": {
            name: "ValidatorError",
            message: "The \"newProps\" parameter in the App component's updateState() function needs to be a an object (but not an array).",
            code: "app-102"
        },
        "app-103": {
            name: "ValidatorError",
            message: "The \"viewProps\" parameter in the App component's handleNavigation() function needs to be an object (but not an array), containing the following keys: \"metaData\" (object), \"viewData\" (object) and \"refreshFactor\" (number).",
            code: "app-103"
        },
        "app-104": {
            name: "ValidatorError",
            message: "The \"viewProps\" parameter in the App component's handleNavigation() function needs to be a an object (but not an array).",
            code: "app-104"
        },
        "app-105": {
            name: "ValidatorError",
            message: "The \"sidebarProps\" parameter in the App component's handleMainNavBarClick() function needs to be a an object (but not an array).",
            code: "app-105"
        },
        "app-106": {
            name: "ValidatorError",
            message: "The \"sidebarProps\" parameter in the App component's handleMainNavBarClick() function is missing an \"activeNavLinkKey\" key (as an integer) in its object.",
            code: "app-106"
        }
    }

    beforeEach(() => {
        jest.clearAllMocks();
        testComponent = predefinedComponent();
        
        componentInstance = testComponent.instance();
        componentInstance.setState = jest.fn();
        componentInstance.launchErrorOverlay = jest.fn();

        ExceptionsHandler.ValidatorError = jest.fn();
        ExceptionsHandler.ValidatorError.mockImplementation(errCode => {
            return actualErrorReturns[errCode];
        });
    });

    describe("Test updateState(newProps, showLoadbar, callback)", () => {
        const various_newProps = [
            ["Test inserting a string"],
            [454],
            [["multidimensional array", "gamma ray bursts", "super novae", 45, false]],
            [true],
            [false],
            [undefined],
            [null],
            [() => {}]
        ];

        const various_newProps_objects = [
            [{testkey: 1, testkey2: "text", testkey3: false}],
            [{testkey4: "randomtext", testkey5: null, testkey6: true}],
            [{testkey7: "magnetar", testkey8: ["gamma", "radio", "nuclear", false], testkey9: false}],
            [{testkey10: "super novae", testkey8: false, testkey9: undefined}]
        ]

        const various_showLoadbar = [
            [true],
            [false],
            ["test inserting another string"],
            [null],
            [undefined],
            [() => {}],
            [{testkey: 1, testkey2: "text", testkey3: false}],
            [["apple", "pearch", "orange", "strawberry"]],
            [77]
        ];

        const various_callback = [
            [() => {}],
            [true],
            [false],
            ["test inserting another string"],
            [null],
            [undefined],
            [{testkey: 7, testkey2: "text", testkey3: true}],
            [["apple", "pearch", "pineapple", "berry"]],
            [241]
        ];

        describe("\"newProps\" (not a non-array object), \"showLoadbar\" (various values), \"callback\" (undefined)", () => {
            for(let i = 0; i < various_showLoadbar.length; i++){
                const testComponent = predefinedComponent();
                const componentInstance = testComponent.instance();

                test.each(various_newProps)("Run updateState(%p, " + various_showLoadbar[i][0] + "): ExceptionsHandler.ValidatorError(\"app-102\") should be called once", (val) => {
                    componentInstance.updateState(val, various_showLoadbar[i][0]);
                    
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-102");
                });
            }

        });

        describe("\"newProps\" (not a non-array object), \"showLoadbar\" (various values), \"callback\" (various values)", () => {
            for(let i = 0; i < various_showLoadbar.length; i++){
                for(let j = 0; j < various_callback.length; j++){
                    const testComponent = predefinedComponent();
                    const componentInstance = testComponent.instance();

                    test.each(various_newProps)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): ExceptionsHandler.ValidatorError(\"app-102\") should be called once.", (val) => {
                        componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                        
                        expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-102");
                    });

                }
            }
        });

        describe("\"newProps\" (non-array object with various keys), \"showLoadbar\" (various values), \"callback\" (various values)", () => {
            const { isUndefined, isFunction } = validator;

            for(let i = 0; i < various_showLoadbar.length; i++){
                for(let j = 0; j < various_callback.length; j++){  
                    const testComponent = predefinedComponent();
                    const componentInstance = testComponent.instance();

                    if(various_showLoadbar[i][0] === true){
                        if(isFunction(various_callback[j][0])){
                            test.each(various_newProps_objects)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): this.setState(newProps, callback) should be called, refreshFactor increased by 1.", (val) => {
                                componentInstance.setState = jest.fn();

                                componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                                val.refreshFactor = componentInstance.state.refreshFactor + 1;

                                expect(componentInstance.setState).toHaveBeenCalledWith(val, various_callback[j][0]);
                            });
                        } else if(isUndefined(various_callback[j][0])){
                            test.each(various_newProps_objects)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): this.setState(newProps) should be called, refreshFactor increased by 1.", (val) => {
                                componentInstance.setState = jest.fn();

                                componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                                val.refreshFactor = componentInstance.state.refreshFactor + 1;

                                expect(componentInstance.setState).toHaveBeenCalledWith(val);
                            });
                        } else {
                            test.each(various_newProps_objects)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): ExceptionsHandler.ValidatorError(\"app-101\") should be called", (val) => {
                                componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                                val.refreshFactor = componentInstance.state.refreshFactor + 1;

                                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-101");
                            });
                        }
                    } else {
                        if(isFunction(various_callback[j][0])){
                            test.each(various_newProps_objects)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): this.setState(newProps, callback) should be called", (val) => {
                                componentInstance.setState = jest.fn();

                                componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                                
                                expect(componentInstance.setState).toHaveBeenCalledWith(val, various_callback[j][0]);
                            });
                        } else if(isUndefined(various_callback[j][0])){
                            test.each(various_newProps_objects)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): this.setState(newProps) should be called", (val) => {
                                componentInstance.setState = jest.fn();

                                componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                                
                                expect(componentInstance.setState).toHaveBeenCalledWith(val);
                            });
                        } else {
                            test.each(various_newProps_objects)("Run updateState(%p, " + various_showLoadbar[i][0] + ", " + various_callback[j][0] + "): ExceptionsHandler.ValidatorError(\"app-101\") should be called", (val) => {
                                componentInstance.updateState(val, various_showLoadbar[i][0], various_callback[j][0]);
                                
                                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-101");
                            });

                        }
                    }
                }
            };
        });

        describe("Check that correct error messages are caught", () => {
            test("Running updateState({}, false, false) will throw an error and pass it to launchErrorOverlay(err), where err.code = \"app-101\"", () => {
                componentInstance.launchErrorOverlay = jest.fn();
                componentInstance.updateState({}, false, false);

                expect(componentInstance.launchErrorOverlay).toHaveBeenCalledWith(expectedErrorReturns["app-101"]);
            })

            test("Running updateState(\"This is definetly not an object\", false, false) will throw an error and pass it to launchErrorOverlay(err), where err.code = \"app-102\"", () => {
                componentInstance.launchErrorOverlay = jest.fn();
                componentInstance.updateState("This is definetly not an object", false, false);

                expect(componentInstance.launchErrorOverlay).toHaveBeenCalledWith(expectedErrorReturns["app-102"]);
            })
        })
    });

    describe("Test handleNavigation(viewProps)", () => {
        describe("\"viewProps\" (not a non-array object)", () => {
            const variousInputs = [
                ["test string"], 
                [30], 
                [true], 
                [false], 
                [["donald duck", "mickey mouse", 5, false]],
                [null],
                [undefined],
            ];
            
            test.each(variousInputs)("Run handleNavigation(%p): ExceptionsHandler.ValidatorError(\"app-104\") should be called", (val) => {
                componentInstance.updateState = jest.fn();
                componentInstance.handleNavigation(val);

                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-104");
            })

            test("Run handleNavigation(): ExceptionsHandler.ValidatorError(\"app-104\") should be called", () => {
                componentInstance.updateState = jest.fn();
                componentInstance.handleNavigation();

                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-104");
            })
        });

        describe("\"viewProps\" (various non-array objects)", () => {
            const various_viewData = [
                ["test string for view data"], 
                [30], 
                [true], 
                [false], 
                [["tiger", "winnie", 5, false]],
                [null],
                [undefined],
                [{testObj1: "contents of test object 1"}],
                [{testObj4: "contents of test object 1", anotherTestObj: 41}]
            ];

            const various_metaData = [
                ["test string for meta data"], 
                [45], 
                [true], 
                [false], 
                [["daisy duck", "goofy", 52, true]],
                [null],
                [undefined],
                [{testObj2: "contents of test object 2"}],
                [{testObj21: "contents of test object 2", anotherTestObj2: false, anotherTestObj3: null}]
            ];

            const various_refreshFactor = [
                ["test string for refreshFactor"], 
                [14], 
                [true], 
                [false], 
                [["darkwing duck", "simba", false, 247]],
                [null],
                [undefined],
                [{testObj3: "contents of test object 3"}],
                [{testObj32: "contents of test object 3", imagination1: 0, anotherTestObj4: "test string", anotherTestObj5: () => {}}]
            ];


            test.each(various_viewData)("Run handleNavigation({ viewData: %p }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", (val) => {
                componentInstance.handleNavigation({ viewData: val });

                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
            });


            test.each(various_metaData)("Run handleNavigation({ metaData: %p }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", (val) => {
                componentInstance.handleNavigation({ metaData: val });

                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
            });

            test.each(various_refreshFactor)("Run handleNavigation({ refreshFactor: %p }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", (val) => {
                componentInstance.handleNavigation({ refreshFactor: val });

                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
            });

            test("Run handleNavigation({}): ExceptionsHandler.ValidatorError(\"app-103\") should be called", () => {
                componentInstance.handleNavigation({});

                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
            });

            describe("\"viewProps\" contains \"viewData\" and \"metaData\" with various values.", () => {
                for(let i = 0; i < various_metaData.length; i++){
                    const testComponent = predefinedComponent();
                    const componentInstance = testComponent.instance();

                    test.each(various_viewData)("Run handleNavigation({ viewData: %p, metaData: " + various_metaData[i][0] + " }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", (val) => {
                        componentInstance.handleNavigation({ viewData: val, metaData: various_metaData[i][0] });

                        expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
                    });

                }
            });

            describe("\"viewProps\" contains \"viewData\" and \"refreshFactor\" with various values.", () => {
                for(let i = 0; i < various_refreshFactor.length; i++){
                    const testComponent = predefinedComponent();
                    const componentInstance = testComponent.instance();

                    test.each(various_viewData)("Run handleNavigation({ viewData: %p, refreshFactor: " + various_refreshFactor[i][0] + " }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", (val) => {
                        componentInstance.handleNavigation({ viewData: val, refreshFactor: various_refreshFactor[i][0] });

                        expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
                    });

                }
            });

            describe("\"viewProps\" contains \"metaData\" and \"refreshFactor\" with various values.", () => {
                for(let i = 0; i < various_refreshFactor.length; i++){
                    const testComponent = predefinedComponent();
                    const componentInstance = testComponent.instance();

                    test.each(various_metaData)("Run handleNavigation({ metaData: %p, refreshFactor: " + various_refreshFactor[i][0] + " }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", (val) => {
                        componentInstance.handleNavigation({ metaData: val, refreshFactor: various_refreshFactor[i][0] });
                      
                        expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
                    });
                }
            });
 
            describe("\"viewProps\" contains \"metaData\", \"viewData\" and \"refreshFactor\" with various values.", () => {
                const { isObject, isNumber } = validator;

                for(let i = 0; i < various_metaData.length; i++){
                    for(let k = 0; k < various_viewData.length; k++){
                        for(let j = 0; j < various_refreshFactor.length; j++){
                            if(!(isObject(various_metaData[i][0]) && isObject(various_viewData[k][0]) && isNumber(various_refreshFactor[j][0]))){
                                const testComponent = predefinedComponent();
                                const componentInstance = testComponent.instance();
                                
                                test("Run handleNavigation({ metaData: " + various_metaData[i][0] + ", viewData: " + various_viewData[k][0] + ", refreshFactor: " + various_refreshFactor[j][0] + " }): ExceptionsHandler.ValidatorError(\"app-103\") should be called", () => {
                                    componentInstance.handleNavigation({ metaData: various_metaData[i][0], viewData: various_viewData[k][0], refreshFactor: various_refreshFactor[j][0] });

                                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-103");
                                });

                            }
                        }
                    }
                }

                const assumedValid_viewProps_inputs = [
                    [{ metaData: {}, viewData: {}, refreshFactor: 21 }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 17 }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 85 }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 61 }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 18 }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 28, randomKey1: {}, randomKey2: () => {}, randomKey3: ["test string"], randomKey4: false }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 42, randomKey1: null, randomKey2: true, randomKey3: ["test string again"], randomKey4: () => {} }],
                    [{ metaData: {}, viewData: {}, refreshFactor: 94, randomKey1: false, randomKey2: [3,2,4], randomKey4: ["test string again"], randomKey3: 248 }]
                ];

                test.each(assumedValid_viewProps_inputs)("Run handleNavigation(%p): this.updateState() should be called", (val) => {
                    const testComponent = predefinedComponent();
                    const componentInstance = testComponent.instance();

                    componentInstance.updateState = jest.fn();
                    componentInstance.handleNavigation(val);

                    expect(componentInstance.updateState).toHaveBeenCalled();
                }); 
            });

        });

        describe("Check that correct error messages are caught", () => {
            test("Running handleNavigation({}) will throw an error and pass it to launchErrorOverlay(err), where err.code = \"app-103\"", () => {
                componentInstance.launchErrorOverlay = jest.fn();
                componentInstance.handleNavigation({});

                expect(componentInstance.launchErrorOverlay).toHaveBeenCalledWith(expectedErrorReturns["app-103"]);
            })

            test("Running handleNavigation(\"This is absolutely not an object\") will throw an error and pass it to launchErrorOverlay(err), where err.code = \"app-104\"", () => {
                componentInstance.launchErrorOverlay = jest.fn();
                componentInstance.handleNavigation("This is absolutely not an object");

                expect(componentInstance.launchErrorOverlay).toHaveBeenCalledWith(expectedErrorReturns["app-104"]);
            })
        }); 
    });

    describe("Test handleMainNavBarClick(sidebarProps)", () => {

        describe("\"sidebarProps\" is not an object", () => {
            const various_sidebarProps = [
                ["A very random input string"],
                [2.47],
                [false],
                [true],
                [null],
                [undefined],
                [[20, 15, "hello"]],
                [() => {}]
            ];

            test("Run handleMainNavBarClick(): ExceptionsHandler.ValidatorError(\"app-105\") should be called", () => {
                componentInstance.handleMainNavBarClick();
    
                expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-105");
            })
        });

        describe("Test when \"sidebarProps\" parameter is an object with various keys.", () => {
            describe("Check that \"sidebarProps\" contains the key \"activeNavLinkKey\" in its first level, and that it contains a number. If not, throw an error", () => {    
                const various_activeNavLinkKey_notNumbers = [
                    ["test string"],
                    [true],
                    [false],
                    [undefined],
                    [null],
                    [{ anotherSetup: "ofObjects", test: true }],
                    [[2,3,4,5]],
                    [() => {}]
                ];

                test.each(various_activeNavLinkKey_notNumbers)("Run handleMainNavBarClick({ activeNavLinkKey: %p}): ExceptionsHandler.ValidatorError(\"app-106\") should be called, because \"activeNavLinkKey\" is not a number in \"sidebarProps\".", (val) => {
                    componentInstance.handleMainNavBarClick({ activeNavLinkKey: val });
        
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-106");
                });

                const various_activeNavLinkKey_numbers = [
                    [77],
                    [48],
                    [120],
                    [5],
                    [9]
                ];

                test.each(various_activeNavLinkKey_numbers)("Run handleMainNavBarClick({ activeNavLinkKey: %p}): componentInstance.updateState should be called, because \"sidebarProps\" contains \"activeNavLinkKey\" with a number in its first level.", (val) => {
                    componentInstance.updateState = jest.fn();
                    componentInstance.handleMainNavBarClick({ activeNavLinkKey: val });
        
                    expect(componentInstance.updateState).toHaveBeenCalled();
                });
            });

            describe("Check that \"sidebarProps\" contains various objects with no valid keys", () => {
                const various_objects_invalid_keys = [
                    [{ key1: "abc", key2: "def", key3: 8, key4: "ghi", key5: "jklmnop" }],
                    [{ key6: false, key7: true, key8: [], key9: undefined, key10: 13 }],
                    [{ key11: 25, key12: false, key13: 8, key14: "ghasdadi", key15: {} }],
                    [{ key16: "abc", key17: "def", key18: 8, key19: "ghi", key20: "jklmnop" }]
                ];

                test.each(various_objects_invalid_keys)("Run handleMainNavBarClick(%p): ExceptionsHandler.ValidatorError(\"app-106\") should be called, because \"activeNavLinkKey\" is missing in the first level of \"sidebarProps\".", (val) => {
                    componentInstance.handleMainNavBarClick(val);
        
                    expect(ExceptionsHandler.ValidatorError).toHaveBeenCalledWith("app-106");
                });
            })
            
        })
        
        describe("Check that correct error messages are caught", () => {
            test("Running handleMainNavBarClick({}) will throw an error and pass it to launchErrorOverlay(err), where err.code = \"app-106\"", () => {
                componentInstance.launchErrorOverlay = jest.fn();
                componentInstance.handleMainNavBarClick({});

                expect(componentInstance.launchErrorOverlay).toHaveBeenCalledWith(expectedErrorReturns["app-106"]);
            })

            test("Running handleMainNavBarClick(\"Not an object\") will throw an error and pass it to launchErrorOverlay(err), where err.code = \"app-105\"", () => {
                componentInstance.launchErrorOverlay = jest.fn();
                componentInstance.handleMainNavBarClick("Not an object");

                expect(componentInstance.launchErrorOverlay).toHaveBeenCalledWith(expectedErrorReturns["app-105"]);
            })
        })
    });
});