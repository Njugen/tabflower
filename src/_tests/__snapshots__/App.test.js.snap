// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <App /> component behaviour at mount Test <App /> render <ViewFooter> is rendered without props 1`] = `
"<Fragment>
  <ErrorBoundary>
    <div className=\\"container-fluid\\">
      <MainNavBar routes={{...}} onMainNavBarClick={[Function: onMainNavBarClick]} />
      <div className=\\"row\\">
        <div className=\\"col-md-12 py-2\\" id=\\"tabeon-view-container\\">
          <RouteList onRaisedRoutesInfo={[Function: onRaisedRoutesInfo]} onRaiseToModal={[Function: onRaiseToModal]} onNavigation={[Function: onNavigation]} onRaiseToErrorOverlay={[Function: onRaiseToErrorOverlay]} />
          <ViewFooter />
        </div>
      </div>
    </div>
    <FullWidthLoadbar refreshFactor={0} />
  </ErrorBoundary>
</Fragment>"
`;
